### PR TITLE
Introduce core library

### PR DESCRIPTION
**Pull request by W95Psp**
This PR:
 - adds `libs` folder with a partial F* `core` model and a few other libraries (see `libs/fstar`);
 - adds a few examples in `examples`;
 - adds a CI for those examples.

### CODE REVIEW

File: .github/workflows/install_and_test.yml [hunk 1]
```diff
@@ -28,6 +28,34 @@ jobs:
...
+
+    - name: Checkout Hacl
+      uses: actions/checkout@v3
+      with:
+        repository: 'hacl-star/hacl-star'
+        path: "hacl-star"
+      
+    - name: Extract and typecheck examples
+      run: |
+        export HACL_HOME="$GITHUB_WORKSPACE/hacl-star"
```
Comment (franziskuskiefer) [inline .github/workflows/install_and_test.yml:43]: We should throw these things into helpers. But this is fine for now.

File: examples/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+[workspace]
+members = [
+     "chacha20",
```
Comment (franziskuskiefer) [inline examples/Cargo.toml:3]: What about kyber-serialize?

File: examples/README.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+| Name     | Status         |
+| -------- | -------------- |
+| chacha20 | Lax-typechecks |
+| lob      | Typechecks     |
+| sha256   | Lax-typechecks |
+
+## How-to extract and typecheck an example
+To generate F* modules and **lax-typecheck** them for an example `<EXAMPLE>`:
+1. `cd <EXAMPLE>/proofs/fstar/extraction`
```
Comment (franziskuskiefer) [inline examples/README.md:11]: The `proofs` directory doesn't exist.
Maybe we should add a script here to extract everything?

File: examples/README.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# Examples
+
+| Name     | Status         |
+| -------- | -------------- |
+| chacha20 | Lax-typechecks |
+| lob      | Typechecks     |
```
Comment (franziskuskiefer) [inline examples/README.md:6]: log fails for me with
```text
* Error 37 at Lob_backend.fst(15,47-15,64):
  - Constructor Candid.Types.Internal.TypeInner_Variant not found
```

File: examples/README.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+| sha256   | Lax-typechecks |
```
Comment (franziskuskiefer) [inline examples/README.md:7]: This fails with
```text
* Error 72 at Sha256.fst(103,44-103,49):
  - Identifier not found: [slice]
```

File: examples/README.md [hunk 6]
```diff
@@ -0,0 +1,13 @@
+# Examples
+
+| Name     | Status         |
+| -------- | -------------- |
+| chacha20 | Lax-typechecks |
```
Comment (franziskuskiefer) [inline examples/README.md:5]: This fails with the same slice error as sha256.

File: examples/chacha20/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    b = update_array(b, plain);
+    b = chacha20_encrypt_block(st0, ctr, &b);
+    b[0..plain.len()].to_vec()
+}
+
+pub fn chacha20_update(st0: State, m: &[u8]) -> Vec<u8> {
+    let mut blocks_out = Vec::new();
+    let num_blocks = m.len() / 64;
+    let remainder_len = m.len() % 64;
+    // for (i, msg_block) in m.chunks(64).enumerate() {
```
Comment (franziskuskiefer) [inline examples/chacha20/src/lib.rs:102]: Let's drop the commented code.

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
Thanks, looks like some things aren't working yet, otherwise this looks good.

File: examples/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,6 @@
...
+     "lob",
+     "sha256",
+]
```
Comment (franziskuskiefer) [inline examples/Cargo.toml:6]: This should get resolver 2.

File: examples/README.md (hunk 3 above)
Comment (franziskuskiefer) [reply examples/README.md:11]: Or just say that `make` runs the extraction as well (which is a little confusing, given that it's run from the extraction dir)

File: examples/Cargo.toml (hunk 2 above)
Comment (W95Psp) [reply examples/Cargo.toml:3]: Yes, we should include it at some point, but for this PR I just want to introduce `core` + a few examples.
Then I'll make a follow-up PR with more examples (and better CI).

**Review by franziskuskiefer — APPROVED** [review]

**Pull request merged**